needy/needy_configuration.py

```diff
@@ -0,0 +1,86 @@
+import os
+import json
+import logging
+
+from .caches.directory import Directory
+from .build_cache import BuildCache
+from .filesystem import os_file, lock_fd
+
+from contextlib import contextmanager
+
+
+class NeedyConfiguration:
+    def __init__(self, base_path):
+        self.__base_path = base_path
+
+        with self.__locked_needyconfigs(self.__base_path) as fds:
+            self.__configuration = self._evaluate_needyconfigs(fds)
+
+    @staticmethod
+    @contextmanager
+    def __locked_needyconfigs(base_path):
+        fds = []
+        path = base_path
+
+        candidates = []
+
+        # Be sure whatever is done here is deadlock free.
+        while path:
+            needyconfig = os.path.join(path, '.needyconfig')
+            candidates.append(needyconfig)
+            if path == os.path.dirname(path):
+                break
+            path = os.path.dirname(path)
+
+        for c in candidates:
+            if os.path.exists(c):
+                logging.debug('using .needyconfig at {}'.format(c))
+
+                f = os_file(c, os.O_RDWR, 'r+')
+                lock_fd(f.fileno())
+                fds.append(f)
```

Comment:
So to restate the goal, we want to make sure that whatever set of configurations we get is actually something that was on the disk at some single point in time.

I don't think this accomplishes that when other files may be created. For externally created files, a solution might be impractical. But I'd like to at least devise a strategy that ensures that files created by other invocations of Needy don't cause race conditions.

`|------|-new1-|-cnf1-|-new2-|-cnf2-|-cdir-|`

Two subsequent Needy commands are invoked to create new1, then new2. How can we ensure that no other Needy process reads a configuration with new2, but not new1?

Furthermore, can we future-proof this so that if a Needy invocation for a different version is running, we reduce the risk of deadlocking due to future changes in locking strategies?

I think the best way to go here might actually be to just lock the config file, creating it if it doesn't exist, in the highest directory that we have permissions for. If it didn't exist, delete it when we're done.